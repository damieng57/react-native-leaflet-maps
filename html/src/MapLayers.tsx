import * as React from "react";
import {
  TileLayer,
  LayersControl,
  WMSTileLayer,
  WMSTileLayerProps,
  TileLayerProps,
  ImageOverlay,
  ImageOverlayProps
} from "react-leaflet";
import { MapLayer, MapLayerType } from "./models";

const { BaseLayer } = LayersControl;

interface MapLayersProps {
  mapLayers: MapLayer[];
}

class MapLayers extends React.Component<MapLayersProps> {
  private Layer = (props: MapLayer): JSX.Element => {
    switch (props.layerType) {
      case MapLayerType.IMAGE_LAYER:
        return <ImageOverlay {...(props as ImageOverlayProps)} />;
      case MapLayerType.WMS_TILE_LAYER:
        return <WMSTileLayer {...(props as WMSTileLayerProps)} />;
      default:
        return <TileLayer {...(props as TileLayerProps)} />;
    }
  };

  private Layers = (): JSX.Element[] => {
    const { mapLayers } = this.props;
    return mapLayers.map(
      (layer: MapLayer, index: number): JSX.Element => {
        if (layer.baseLayerName && mapLayers.length > 1) {
          return (
            <BaseLayer
              key={`layer-${index}`}
              checked={layer.baseLayerIsChecked ?? false}
              name={layer.baseLayerName || `Layer.${index}`}
            >
              <this.Layer {...layer} />
            </BaseLayer>
          );
        }
        return <this.Layer key={`layer-${index}`} {...layer} />;
      }
    );
  };

  render() {
    const { mapLayers } = this.props;
    if (mapLayers.length > 1) {
      return <LayersControl>{this.Layers()}</LayersControl>;
    } else {
      return <>{this.Layers()}</>;
    }
  }
}

export default MapLayers;
