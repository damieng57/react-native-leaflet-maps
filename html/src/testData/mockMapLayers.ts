import { MapLayer, MapLayerType } from "../models";

const mockMapLayers: MapLayer[] = [
  {
    attribution:
      '&amp;copy <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
    baseLayerIsChecked: true,
    baseLayerName: "OpenStreetMap.Mapnik",
    url: "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
  },
  {
    attribution:
      '&amp;copy <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
    baseLayerIsChecked: false,
    baseLayerName: "OpenStreetMap.BlackAndWhite",
    url: "https://tiles.wmflabs.org/bw-mapnik/{z}/{x}/{y}.png"
  },
  {
    baseLayerName: "WMS Tile Layer",
    subLayer: "nasa:bluemarble",
    layerType: MapLayerType.WMS_TILE_LAYER,
    url: "https://demo.boundlessgeo.com/geoserver/ows"
  },
  {
    baseLayerName: "Image",
    layerType: MapLayerType.IMAGE_LAYER,
    url: "http://www.lib.utexas.edu/maps/historical/newark_nj_1922.jpg",
    bounds: [
      [40.712216, -74.22655],
      [40.773941, -74.12544]
    ]
  }
];

export default mockMapLayers;
