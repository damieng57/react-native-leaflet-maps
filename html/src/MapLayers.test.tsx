import * as React from "react";
import { default as MapLayerComponent } from "./MapLayers";
import { render } from "@testing-library/react";
import mockMapLayers from "./testData/mockMapLayers";
import { Map } from "react-leaflet";

describe("MapLayers Component", () => {
  test("it renders", () => {
    console.log(MapLayerComponent);
    const { asFragment } = render(
      <Map center={[0, 0]} zoom={13} style={{ width: 100, height: 100 }}>
        <MapLayerComponent mapLayers={mockMapLayers} />
      </Map>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
