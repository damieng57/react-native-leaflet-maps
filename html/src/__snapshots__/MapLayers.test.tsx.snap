// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapLayers Component it renders 1`] = `
<DocumentFragment>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    style="width: 100px; height: 100px; position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://a.tile.openstreetmap.org/13/4095/4095.png"
              style="width: 256px; height: 256px; left: -256px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://b.tile.openstreetmap.org/13/4096/4095.png"
              style="width: 256px; height: 256px; left: 0px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://b.tile.openstreetmap.org/13/4095/4096.png"
              style="width: 256px; height: 256px; left: -256px; top: 0px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://c.tile.openstreetmap.org/13/4096/4096.png"
              style="width: 256px; height: 256px; left: 0px; top: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            âˆ’
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      >
        <div
          aria-haspopup="true"
          class="leaflet-control-layers leaflet-control"
        >
          <a
            class="leaflet-control-layers-toggle"
            href="#"
            title="Layers"
          />
          <section
            class="leaflet-control-layers-list"
          >
            <div
              class="leaflet-control-layers-base"
            >
              <label>
                <div>
                  <input
                    checked="checked"
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_33"
                    type="radio"
                  />
                  <span>
                     OpenStreetMap.Mapnik
                  </span>
                </div>
              </label>
              <label>
                <div>
                  <input
                    class="leaflet-control-layers-selector"
                    name="leaflet-base-layers_33"
                    type="radio"
                  />
                  <span>
                     OpenStreetMap.BlackAndWhite
                  </span>
                </div>
              </label>
            </div>
            <div
              class="leaflet-control-layers-separator"
              style="display: none;"
            />
            <div
              class="leaflet-control-layers-overlays"
            />
          </section>
        </div>
      </div>
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
           | &copy 
          <a
            href="http://osm.org/copyright"
          >
            OpenStreetMap
          </a>
           contributors
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
